src/backend/cdb/motion/cdbmotion.c

```diff
@@ -1274,3 +1277,143 @@ UpdateSentRecordCache(int32 *sent_record_typmod)
+	{
+		if (!pg_strcasecmp(IPCLayerImpls[i]->type_name, type_name))
+		{
+			CurrentMotionIPCLayer = IPCLayerImpls[i];
+			Gp_interconnect_type  = CurrentMotionIPCLayer->ic_type;
+			return;
+		}
+	}
+
+	/* should never run here */
+	ereport(ERROR,
+			(errcode(ERRCODE_GP_INTERCONNECTION_ERROR),
+			 errmsg("No IPC layer implement found with type: \"%s\".",
+					type_name)));
+
+	return;
+}
+
+/*
+ * Check the new value of gp_interconnect_type with type_name of the loaded IPC
+ * layer implements.
+ */
+bool
+CheckGpInterconnectTypeStr(char **type_name)
+{
+	/*
+	 * do nothing before interconnect.so loaded.
+	 *
+	 * gp_interconnect_type will be checked after interconnect.so is loaded
+	 * in PostmasterMain() by calling ResetCurrentMotionIPCLayer().
+	 */
+	if (!process_shared_preload_libraries_done)
+		return true;
+
+	for (int i = 0; i < CurrentIPCLayerImplNum; ++i)
+	{
+		if (!pg_strcasecmp(IPCLayerImpls[i]->type_name, *type_name))
+			return true;
+	}
+
+	return false;
+}
+
+/*
+ * Called by interconnect.so to register a new IPC layer implement.
+ */
+void
+RegisterIPCLayerImpl(MotionIPCLayer *impl)
+{
+	if (CurrentIPCLayerImplNum >= MAX_NUMBER_TYPES)
+	{
+		ereport(ERROR,
+				(errcode(ERRCODE_GP_INTERCONNECTION_ERROR),
+				 errmsg("There is no free entry for a new IPC layer implement.")));
+
+		return;
+	}
+
+	for (int i = 0; i < CurrentIPCLayerImplNum; ++i)
+	{
+		if (impl->ic_type == IPCLayerImpls[i]->ic_type)
+		{
+			ereport(ERROR,
+					(errcode(ERRCODE_GP_INTERCONNECTION_ERROR),
+					 errmsg("type: \"%s\" has been registered.", impl->type_name)));
+
+			return;
+		}
+	}
+
+	IPCLayerImpls[CurrentIPCLayerImplNum++] = impl;
+}
+
+/*
+ * Called by PostmasterMain() to reset the CurrentMotionIPCLayer after
+ * interconnect.so is loaded.
+ */
+void
+ResetCurrentMotionIPCLayer(void)
+{
+	const char *cur_val;
+	const char *reset_val;
+	StringInfoData types;
+
+	/*
+	 * Do nothing if no any IPC layer implement loaded, such as singlenode.
+	 */
+	if (CurrentIPCLayerImplNum <= 0)
+		return;
+
+	/*
+	 * Get current value and reset_val of gp_interconnect_type.
+	 *
+	 * NOTE: called by PostmasterMain() after interconnect.so is loaded,
+	 *       cur_val and reset_val should be the same.
+	 */
+	cur_val   = GetConfigOption("gp_interconnect_type", false, false);
+	reset_val = GetConfigOptionResetString("gp_interconnect_type");
+
+	Assert(cur_val);
+	Assert(reset_val);
+	Assert(pg_strcasecmp(cur_val, reset_val) == 0);
+
+	/*
+	 * Check cur_val by loaded IPC layer implements.
+	 */
+	for (int i = 0; i < CurrentIPCLayerImplNum; ++i)
+	{
+		if (!pg_strcasecmp(cur_val, IPCLayerImpls[i]->type_name))
+		{
+			SetCurrentMotionIPCLayer(IPCLayerImpls[i]->type_name);
+			return;
+		}
+	}
+
+	/*
+	 * Report error with valid types.
+	 */
+	initStringInfo(&types);
+	for (int i = 0; i < CurrentIPCLayerImplNum; ++i)
+		appendStringInfo(&types, "%s, ", IPCLayerImpls[i]->type_name);
+	types.data[types.len - 2] = '\0';
```

Comment:
It's better not remove trailing comma like this. The assumption `CurrentIPCLayerImplNum > 0` is placed far from here.

Reply:
fix in [commit](https://github.com/cloudberrydb/cloudberrydb/pull/646/commits/82b281fbea95edb8545b90632e8cf034fb6eb463)